src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Serialization/ObjectReaders.cs

```diff
@@ -189,20 +189,18 @@ private static void ProcessMetadata(JsonDataReader reader, IDictionary<string, s
 
     public static ProjectRazorJson ReadProjectRazorJsonFromProperties(JsonDataReader reader)
     {
-        ProjectRazorJsonData data = default;
-        reader.ReadProperties(ref data, ProjectRazorJsonData.PropertyMap);
-
-        // We need to add a serialization format to the project response to indicate that this version
-        // of the code is compatible with what's being serialized. This scenario typically happens when
-        // a user has an incompatible serialized project snapshot but is using the latest Razor bits.
-
-        if (string.IsNullOrEmpty(data.SerializationFormat) || data.SerializationFormat != ProjectSerializationFormat.Version)
+        if (!reader.TryReadInt32(WellKnownPropertyNames.Version, out var version) || version != ProjectRazorJson.Version)
         {
-            // Unknown serialization format.
-            return null!;
+            throw new ProjectRazorJsonSerializationException(SR.Unsupported_project_razor_json_version_encountered);
```

Comment:
I'm curious, what is the experience like in the IDE when this exception is thrown. Does everything work gracefully until an updated json file is written?

Reply:
Honestly, it's just fine. The exception gets caught here:

https://github.com/dotnet/razor/blob/fd4a14b7f8ae27795827c3c6bab71accaafa2831/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Serialization/JsonFileDeserializer.cs#L26-L40

The IDE seems to just treat it as if the project is newly created and overwrites the existing project.razor.json file.

Reply:
Glad to hear it, otherwise I was going to suggest some kind of semver where we could reason about breaking changes versus changes where it was okay to limp along with defaults for some things etc. 😛

Reply:
I think I'm sensing some of your trademark sarcasm. :smile: